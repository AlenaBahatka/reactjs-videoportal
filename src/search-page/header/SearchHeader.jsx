import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import FilterOption from '../common-components/FilterOption';

class SearchHeader extends PureComponent {
	constructor(props) {
		super(props);
		this.state = {
			filterOption: props.filterOptions.defaultChecked,
			query: ''
		};
		this.handleOptionChange = this.handleOptionChange.bind(this);
		this.handleSearchClick = this.handleSearchClick.bind(this);
		this.handleInputChange = this.handleInputChange.bind(this);
		this.keyPress = this.keyPress.bind(this);
	}

	handleSearchClick() {
		let { query, filterOption } = this.state;
		if (query) {
			this.props.handleSearchClick(query, filterOption);
		} else {
			alert('check your query');
		}
		// TODO: remove later. just for logging
		console.log('run search with filter =', this.state.filterOption, '& search query = ', this.state.query);
	}

	handleOptionChange(checkedOption) {
		this.setState({
			filterOption: checkedOption
		});
	}
	handleInputChange(event) {
		this.setState({
			query: event.target.value
		});
	}

	keyPress(event) {
		if (event.keyCode == 13) {
			this.handleInputChange(event);
			this.handleSearchClick();
		}
	}

	render() {
		let { filterOptions } = this.props.filterOptions;
		return (
			<div>
				<div className="container">
					<h2> Find your movie </h2>
					<input
						id="searchInput"
						onKeyDown={this.keyPress}
						type="text"
						value={this.state.query}
						onChange={this.handleInputChange}
					/>
				</div>
				<div className="row filterList">
					<div id="filterOptionsContainer">
						<span>Search by</span>
						<ul>
							{filterOptions.map((option, i) => (
								<li key={option.name}>
									<FilterOption
										checkedOption={this.state.filterOption}
										optionName="filter"
										name={option.name}
										displayName={option.name}
										handleOptionChange={this.handleOptionChange}
									/>
								</li>
							))}
						</ul>
					</div>
					<div>
						<button id="searchButton" className="btn" onClick={this.handleSearchClick}>
							Search
						</button>
					</div>
				</div>
			</div>
		);
	}
}

SearchHeader.propTypes = {
	filterOptions: PropTypes.object
};

export default SearchHeader;
