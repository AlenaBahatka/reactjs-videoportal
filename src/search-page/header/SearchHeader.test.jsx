import React from 'react';
import { shallow, mount } from 'enzyme';
import SearchHeader from './SearchHeader';

let searchHeader, handleSearchClickMock, searchHeaderComponent;

describe('SearchHeader', () => {
	beforeEach(() => {
		handleSearchClickMock = jest.fn();
		let filterOptions = {
			filterOptions: [ { name: 'title' }, { name: 'genre' } ],
			defaultChecked: 'title'
		};
		searchHeader = <SearchHeader filterOptions={filterOptions} handleSearchClick={handleSearchClickMock} />;
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('should correctly render component', () => {
		searchHeaderComponent = shallow(searchHeader);

		expect(searchHeaderComponent).toMatchSnapshot();
	});

	it('should call handleSearchClick on button click in case query was set', () => {
		searchHeaderComponent = mount(searchHeader);
		searchHeaderComponent.setState({ query: 'a' });
		searchHeaderComponent.find('#searchButton').simulate('click');

		expect(handleSearchClickMock).toBeCalled();
	});

	it('should not call handleSearchClick on button click in case query was not set', () => {
		searchHeaderComponent = mount(searchHeader);
		searchHeaderComponent.find('#searchButton').simulate('click');

		expect(handleSearchClickMock).not.toBeCalled();
	});

	it('should change state on input change', () => {
		const expectedNewValue = 'My new value';
		searchHeaderComponent = mount(searchHeader);
		searchHeaderComponent.find('#searchInput').simulate('change', { target: { value: expectedNewValue } });

		expect(searchHeaderComponent.state().query).toEqual(expectedNewValue);
	});

	it('should call some methods in case Enter was pressed', () => {
		const handleInputChangeSpy = jest.spyOn(SearchHeader.prototype, 'handleInputChange');
		const handleSearchClickSpy = jest.spyOn(SearchHeader.prototype, 'handleSearchClick');
		searchHeaderComponent = mount(searchHeader);

		searchHeaderComponent.find('#searchInput').simulate('keyDown', { key: 'Enter', keyCode: 13, which: 13 });

		expect(handleInputChangeSpy).toBeCalled();
		expect(handleSearchClickSpy).toBeCalled();
	});

	it('should call some methods in case Enter was pressed', () => {
		const handleInputChangeSpy = jest.spyOn(SearchHeader.prototype, 'handleInputChange');
		const handleSearchClickSpy = jest.spyOn(SearchHeader.prototype, 'handleSearchClick');
		searchHeaderComponent = mount(searchHeader);

		searchHeaderComponent.find('#searchInput').simulate('keyDown', { key: 'Escape', keyCode: 27, which: '27' });

		expect(handleInputChangeSpy).not.toBeCalled();
		expect(handleSearchClickSpy).not.toBeCalled();
	});

	it('should call handleOptionChange', () => {
		const handleOptionChangeSpy = jest.spyOn(SearchHeader.prototype, 'handleOptionChange');
		searchHeaderComponent = mount(searchHeader);

		searchHeaderComponent.find("input[value='genre']").simulate('change');

		expect(handleOptionChangeSpy).toBeCalled();
	});
});
