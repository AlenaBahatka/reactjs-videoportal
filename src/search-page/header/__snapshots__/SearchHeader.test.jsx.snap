// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchHeader should correctly render component 1`] = `
<div>
  <div
    className="container"
  >
    <h2>
       Find your movie 
    </h2>
    <input
      id="searchInput"
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="row filterList"
  >
    <div
      id="filterOptionsContainer"
    >
      <span>
        Search by
      </span>
      <ul>
        <li
          key="title"
        >
          <FilterOption
            checkedOption="title"
            displayName="title"
            handleOptionChange={[Function]}
            name="title"
            optionName="filter"
          />
        </li>
        <li
          key="genre"
        >
          <FilterOption
            checkedOption="title"
            displayName="genre"
            handleOptionChange={[Function]}
            name="genre"
            optionName="filter"
          />
        </li>
      </ul>
    </div>
    <div>
      <button
        className="btn"
        id="searchButton"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
  </div>
</div>
`;
